### PR TITLE
Update failed validation message to allow a specify email for package signing failure

### PR DESCRIPTION
**Pull request by joelverhagen**
Addresses https://github.com/NuGet/NuGetGallery/issues/5194.

These APIs will be consumed by the orchestrator.

### CODE REVIEW

File: src/NuGetGallery.Core/Services/CoreMessageService.cs [hunk 1]
```diff
@@ -62,16 +62,41 @@ public void SendPackageAddedNotice(Package package, string packageUrl, string pa
...
         {
             string subject = "[{0}] Package validation failed - {1} {2}";
-            string body = @"The package [{1} {2}]({3}) failed validation and was therefore not published on {0}. Note that the package will not be available for consumption and you will not be able to push the same package ID and version until further action is taken. Please [contact support]({4}) for next steps.
+            string body = @"The package [{1} {2}]({3}) failed validation and was therefore not published on {0}. Note that the package will not be available for consumption and you will not be able to push the same package ID and version until further action is taken. Please [contact support]({4}) for next steps.";
 
------------------------------------------------
-<em style=""font-size: 0.8em;"">
-    To stop receiving emails as an owner of this package, sign in to the {0} and
-    [change your email notification settings]({5}).
-</em>";
```
Comment (chenriksson) [inline src/NuGetGallery.Core/Services/CoreMessageService.cs]: Why remove this? Seems like something we should have on all notification emails - a way to unsubscribe.

**Comment by karann-msft** [comment]
This PR has changes for the email we send, not the message we display on the package details page. Correct?

File: src/NuGetGallery.Core/Services/CoreMessageService.cs [hunk 2]
```diff
@@ -100,6 +125,14 @@ public void SendPackageValidationFailedNotice(Package package, string packageUrl
             }
         }
 
+        protected static void AddAllOwnersToMailMessage(PackageRegistration packageRegistration, MailMessage mailMessage)
+        {
+            foreach (var owner in packageRegistration.Owners)
+            {
+                mailMessage.To.Add(owner.ToMailAddress());
```
Comment (chenriksson) [inline src/NuGetGallery.Core/Services/CoreMessageService.cs]: Should we send separate mails, or BCC owners?

File: src/NuGetGallery.Core/Services/CoreMessageService.cs (hunk 1 above)
Comment (karann-msft) [reply src/NuGetGallery.Core/Services/CoreMessageService.cs]: You can't unsubscribe from failure related emails at present.

File: src/NuGetGallery.Core/Services/CoreMessageService.cs [hunk 3]
```diff
@@ -62,16 +62,41 @@ public void SendPackageAddedNotice(Package package, string packageUrl, string pa
...
+                mailMessage.Body = body;
+                mailMessage.From = CoreConfiguration.GalleryNoReplyAddress;
+
+                AddAllOwnersToMailMessage(package.PackageRegistration, mailMessage);
+
+                if (mailMessage.To.Any())
+                {
+                    SendMessage(mailMessage, copySender: false);
+                }
+            }
```
Comment (shishirx34) [inline src/NuGetGallery.Core/Services/CoreMessageService.cs]: lines 81-93 can be refactored into a separate method and reused in the function below `SendPackageSignedFailureNotice`

File: src/NuGetGallery.Core/Services/CoreMessageService.cs [hunk 4]
```diff
@@ -62,16 +62,41 @@ public void SendPackageAddedNotice(Package package, string packageUrl, string pa
...
+        }
+
+        public void SendPackageSignedFailureNotice(Package package, string packageUrl, string announcementsUrl, string twitterUrl)
```
Comment (shishirx34) [inline src/NuGetGallery.Core/Services/CoreMessageService.cs]: The name suggests that there was a failure to sign the package, perhaps `SendSignedPackageNotAllowedNotice` be better or something else you can think of?

File: tests/NuGetGallery.Facts/Services/MessageServiceFacts.cs [hunk 5]
```diff
@@ -865,17 +864,25 @@ public void WillSendEmailToAllOwners(string version)
...
             }
 
-            [Fact]
-            public void WillNotSendEmailToOwnerThatOptsOut()
+            public static IEnumerable<object[]> EmailSettingsCombinations
+                => from u1pa in new[] { true, false }
+                   from u2pa in new[] { true, false }
+                   from u1ea in new[] { true, false }
+                   from u2ea in new[] { true, false }
+                   select new object[] { u1pa, u2pa, u1ea, u2ea };
```
Comment (shishirx34) [inline tests/NuGetGallery.Facts/Services/MessageServiceFacts.cs]: what does `u` `p` `a` mean? can the names be more descriptive of the combination?

Comment (shishirx34) [reply tests/NuGetGallery.Facts/Services/MessageServiceFacts.cs]: I see `user` `push` `allowed` `email`.

**Comment by joelverhagen** [comment]
> This PR has changes for the email we send, not the message we display on the package details page. Correct?

Correct.

File: src/NuGetGallery.Core/Services/CoreMessageService.cs (hunk 2 above)
Comment (joelverhagen) [reply src/NuGetGallery.Core/Services/CoreMessageService.cs]: Out of scope. https://github.com/NuGet/Engineering/issues/1059

File: src/NuGetGallery.Core/Services/CoreMessageService.cs (hunk 3 above)
Comment (joelverhagen) [reply src/NuGetGallery.Core/Services/CoreMessageService.cs]: I tried this but found the number of parameters of this generic helper method grew to a mess (after looking at `MessageService` with a lot of variants.

File: src/NuGetGallery.Core/Services/CoreMessageService.cs (hunk 4 above)
Comment (joelverhagen) [reply src/NuGetGallery.Core/Services/CoreMessageService.cs]: Yeah, I like that.

File: tests/NuGetGallery.Facts/Services/MessageServiceFacts.cs (hunk 5 above)
Comment (joelverhagen) [reply tests/NuGetGallery.Facts/Services/MessageServiceFacts.cs]: Yup.

**Pull request merged**